File: tools/asset_extractor/assets/animation.py
from assets.base import BaseAsset, Reader, opt_param

class Animation(BaseAsset):
    def __init__(self, path: str, addr: int, size: int, options: any) -> None:
        super().__init__(path, addr, size, options)

    def extract_binary(self, rom: bytearray) -> None:
        reader = Reader(rom[self.addr:self.addr+self.size])
        lines = []
        end_of_animation = False
        while not end_of_animation and reader.cursor+3 < self.size:
            frame_index = reader.read_u8()
            keyframe_duration = reader.read_u8()
            bitfield = reader.read_u8()
            bitfield2 = reader.read_u8()

            end_of_animation = bitfield2 & 0x80 != 0
            line = f'\tkeyframe frame_index={frame_index}'
            line += opt_param('duration', '0', str(keyframe_duration))
            line += opt_param('bitfield', '0x0', hex(bitfield))
            line += opt_param('bitfield2', '0x0', hex(bitfield2))
            lines.append(line + '\n')
        if not end_of_animation:
            lines.append('@ TODO why no terminator?\n')
        while reader.cursor < self.size:
            keyframe_count = reader.read_u8()
            lines.append(f'\t.byte {keyframe_count} @ keyframe count\n')

        assert(self.path.endswith('.bin'))
        path = self.path[0:-4] + '.s'
        with open(path, 'w') as file:
            file.writelines(lines)

File: tools/src/asset_processor/assets/animation.cpp
#include "animation.h"
#include "reader.h"
#include <fmt/format.h>
#include <util/file.h>

void AnimationAsset::convertToHumanReadable(const std::vector<char>& baserom) {
    Reader reader(baserom, start, size);
    bool end_of_animation = false;
    auto file = util::open_file(assetPath, "w");
    while (!end_of_animation && reader.cursor + 3 < size) {
        u8 frame_index = reader.read_u8();
        u8 keyframe_duration = reader.read_u8();
        u8 bitfield = reader.read_u8();
        u8 bitfield2 = reader.read_u8();
        end_of_animation = (bitfield2 & 0x80) != 0;
        auto line = fmt::format("\tkeyframe frame_index={}", frame_index);
        line += opt_param(", duration={}", 0, keyframe_duration);
        line += opt_param(", bitfield={:#x}", 0, bitfield);
        line += opt_param(", bitfield2={:#x}", 0, bitfield2);
        std::fputs(line.c_str(), file.get());
        std::fputc('\n', file.get());
    }

    if (!end_of_animation) {
        std::fputs("@ TODO why no terminator?\n", file.get());
    }

    while (reader.cursor < size) {
        u8 keyframe_count = reader.read_u8();
        fmt::print(file.get(), "\t.byte {} @ keyframe count\n", keyframe_count);
    }
}
